pkg/server/deploy/config.go

```diff
@@ -85,28 +92,26 @@ func getDeployConfigFilesFromTarBall(tarBall io.ReadSeeker) (*DeployConfigFiles,
 			return nil, err
 		}
 
-		if hdr.Name != TeresaYamlFileName && hdr.Name != ProcfileFileName {
+		if hdr.Name != tYamlFilename && hdr.Name != tYamlProcessTypeFileName && hdr.Name != ProcfileFileName {
 			continue
 		}
 
-		if hdr.Name == TeresaYamlFileName {
-			deployFiles.TeresaYaml = new(TeresaYaml)
-			if err := readFileFromTarBall(tarReader, deployFiles.TeresaYaml); err != nil {
-				return nil, err
-			}
-			if err := validateTeresaYaml(deployFiles.TeresaYaml); err != nil {
-				return nil, err
+		if hdr.Name == tYamlFilename || hdr.Name == tYamlProcessTypeFileName {
+			if hdr.Name == tYamlProcessTypeFileName {
+				if err := deployFiles.fillTeresaYaml(tarReader); err != nil {
+					return nil, err
+				}
+			} else if deployFiles.TeresaYaml == nil {
+				if err := deployFiles.fillTeresaYaml(tarReader); err != nil {
+					return nil, err
+				}
 			}
 		} else {
 			deployFiles.Procfile = make(map[string]string)
 			if err := readFileFromTarBall(tarReader, deployFiles.Procfile); err != nil {
 				return nil, err
 			}
 		}
```

Comment:
Why you removed that last if? (`if deployFiles.TeresaYaml != nil && deployFiles.Procfile != nil`).
That "if" will prevent Teresa to read a bunch of files unnecessary, or am I miss something?